File: bench/comparison/tdzdd/matching.cpp
// Perfect matching counting via TdZdd's DegreeConstraint: every vertex constrained to
// degree exactly 1. Compared against ZDD.Net's MatchingSpec(perfect: true)
// (PerfectMatching_Grid6x6) and Graphillion's GraphSet.perfect_matchings()
// (matching.py). See ../README.md.
//
// Usage: ./matching gridN.dat
#include <tdzdd/DdStructure.hpp>
#include <tdzdd/spec/DegreeConstraint.hpp>
#include <tdzdd/util/Graph.hpp>
#include <tdzdd/util/IntSubset.hpp>

#include "common.hpp"

using namespace tdzdd;

int main(int argc, char** argv) {
    if (argc != 2) {
        std::fprintf(stderr, "usage: %s <grid.dat>\n", argv[0]);
        return 1;
    }

    Graph graph;
    graph.readAdjacencyList(argv[1]);

    bench::Timer timer;
    IntRange exactlyOne(1, 1);
    DegreeConstraint dc(graph, &exactlyOne);
    DdStructure<2> f(dc);
    f.zddReduce();
    double elapsedMs = timer.elapsedMs();

    bench::report(argv[1], elapsedMs, f.evaluate(ZddCardinality<>()), f.size());
    return 0;
}

File: bench/comparison/graphillion/matching.py
"""Perfect matching counting via Graphillion's GraphSet.perfect_matchings(), on an
n x n grid — matches ZDD.Net's MatchingSpec(perfect: true) (PerfectMatching_Grid6x6) and
TdZdd's DegreeConstraint (matching.cpp). See ../README.md.

Usage: python3 matching.py <n>
"""
import sys

from common import Timer, report
from graphillion import GraphSet
from grid_paths import grid_edges


def run(n):
    edges = grid_edges(n)
    GraphSet.set_universe(edges)

    timer = Timer()
    gs = GraphSet.perfect_matchings()
    count = gs.len()
    elapsed_ms = timer.elapsed_ms()

    report(f"PerfectMatching_Grid{n}x{n}", elapsed_ms, count)


if __name__ == "__main__":
    for arg in sys.argv[1:]:
        run(int(arg))
